src/lib/message_filters.py

```diff
@@ -124,3 +125,47 @@ def filter_noise(messages, message_key, noise_fn):
         log.info(f"Filtered out messages identified as noise. "
                  f"Returning {len(filtered)}/{len(messages)} messages.")
         return filtered
+
+    @staticmethod
+    def filter_katikati_sms_survey_messages(messages, time_keys):
+        """
+        Filters out a list of messages received during Katikati sms survey(s) broadcast period.
+
+        :param messages: List of message objects to filter.
+        :type messages: list of TracedData
+        :param time_keys: Keys in each TracedData object that contain the time the message was sent.
+                          Each TracedData should have exactly one match for each key.
+                          The values must be strings in ISO 8601 format.
+        :type time_keys: set of str
+        :return: Filtered list.
+        :rtype: list of TracedData
+        """
+        # De-duplicate time_keys
+        assert isinstance(time_keys, set)
+
+        log.debug(f"Filtering out messages received during katikati sms survey broadcast period"
+                  f"for time keys {time_keys}...")
+
+        # Not all episodes have Katikati Sms Survey check those that have and filter the messages
+        katikati_survey_messages = []
+        for episode_plan in PipelineConfiguration.RQA_CODING_PLANS:
+            if episode_plan.katikati_survey_start_time is None:
+                continue
+
+            episode_katikati_survey_messages = 0
+            for td in messages:
+                if episode_plan.raw_field in td:
+                    for time_key in time_keys:
+                        if time_key in td and isoparse(episode_plan.katikati_survey_start_time) <= isoparse(td[time_key]) \
```

Comment:
I think you should do a check that each td has only one of the time keys, otherwise you could double-count messages

Reply:
filter_time_range function running before validates the input data to ensure that each message object only contains one of the time_keys. Not sure if there is a need to check again.

Reply:
The problem is filter_time_range isn't called from here, so you're relying on all users of this function to remember to call filter_time_range first. So I think it would be safer to do that check in this function too.

Reply:
That's right. updated thanks